### PR TITLE
Implement EEGLAB-style interpolation in Reference for matlab_strict

### LINKED ISSUE #77: Interpolation of bad channels differs numerically between PyPREP and MatPREP
For whatever reason, MATLAB PREP uses a different method of interpolation for channel prediction during RANSAC than it does for interpolating channels during and after re-referencing.

For RANSAC, they use their own internal interpolation code (a copy of the private interpolation code from EEGLAB's `clean_rawdata` plugin), which happens to match MNE's `_make_interpolation_matrix` exactly. For all other interpolation, they use a wrapper around EEGLAB's public interpolation code, which seems to use a different method of spherical spline interpolation (different numeric results, no reference to Perrin '89 in the docstring).

Anyway, I can't think of any a priori justification for using different interpolation logic in one place over the other and my gut is to trust MNE's code more than EEGLAB's, but regardless it's something that needs to be fixed for `Reference` to produce numerically-identical results to MATLAB PREP for #58. According to the original manuscript it seems like they comparison-tested the methods and found they gave similar correlations with actual signals?

> The PREP pipeline uses the `spherical` option of EEGLAB `eeg_interp` function for channel interpolation. This function uses Legendre polynomials up through degree 7. To test this choice, we compared this interpolation method with the `v4` option and two other spherical interpolation functions, including the one used in RANSAC. **We applied a rank sum test on the correlation of the interpolated channel with the actual channel and found no significant difference between the three spherical interpolation methods**. The `v4` option performed significantly worse. The results were not sensitive to whether a block of channels was removed in the immediate vicinity of the bad channel or whether the interpolation was done in ICA space and projected back.

Sort of confused by "there were no significant differences, but `v4` was significantly worse", though. May be worth doing our own testing.

### PR DESCRIPTION
**Pull request by a-hurst**
# PR Description

Closes #77. This PR basically reimplements most of EEGLAB's `eeg_interp` method in Python so we can finally do MATLAB equivalence tests for the entirety of the `Reference` class (also added in this PR). This was... *not easy* to say the least, but it was necessary for the peace of mind that these MatPREP comparison CI tests allow (and also helped us identify a bunch of other small bugs/discrepancies along the way).

This is also the last numeric difference from MatPREP I'm aware of other than adaptive line noise removal (and differences between EEGLAB's montage-importing code and MNE's which result in differences in RANSAC, but that's not exactly a *PREP* divergence as much as an EEGLAB/MNE one).

Anyway, I've tried my best to make sense of and comment/document the re-implemented EEGLAB logic, but there are still a few areas I don't fully understand. If you have any insights, please let me know!

# Merge Checklist

- [x] the PR has been reviewed and all comments are resolved
- [x] all [CI][what-is-ci] checks pass
- [x] (if applicable): the PR description includes the phrase `closes #<issue-number>` to [automatically close an issue][auto-close-documentation]
- [x] (if applicable): bug fixes, new features, or [API][what-is-api] changes are documented in [whats_new.rst][whats-new-file]

[what-is-ci]: https://help.github.com/en/actions/building-and-testing-code-with-continuous-integration/about-continuous-integration
[auto-close-documentation]: https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
[what-is-api]: https://en.wikipedia.org/wiki/Application_programming_interface
[whats-new-file]: https://github.com/sappelhoff/pyprep/blob/master/docs/whats_new.rst

### CODE REVIEW

**Comment by a-hurst** [comment]
Note: if the docstring for `_eeglab_calc_g` seems a little opaque, that's because it's more-or-less a direct copy of the docstring for MNE's [equivalent internal function](https://github.com/mne-tools/mne-python/blob/3078533f8704820a168797578f5b47ddff4be347/mne/channels/interpolation.py#L35-L52). I'm not sure I understand the logic behind the actual math well enough to write anything more informative.

File: docs/matlab_differences.rst
Comment (sappelhoff) [inline docs/matlab_differences.rst:214]:
```suggestion
when the ``matlab_strict`` parameter is set to ``True``.
```

File: tests/test_matprep_compare.py
Comment (sappelhoff) [inline tests/test_matprep_compare.py:364]: Can you open an issue for that once this PR is merged? That'd make it easier to track, as code comments are easily forgotten

**Review by sappelhoff — COMMENTED** [review]
Thanks! Nice to see all the equivalence tests :-)

One more general comments about the EEGLAB "translation" to Python --> i'd appreciate if you could go to the EEGLAB GitHub repo and create "permanent links" to the parts of the codebase that you used for the translation.

For example, you have these two lines of code:

```python
  # Subtract all the Euclidian electrode distances from 1 (why?)
    EI = np.ones([n_to, n_from]) - elec_dists
```

This would be significantly enhanced (from a dev/docs perspective) if you add one more comment line like this (note, that's just an example):

```python
# https://github.com/sccn/eeglab/blob/167dfc8766a253619fc586848c704665e39bce9f/Contents.m#L8
```

You can also shorten the hash to 7 characters, instead of using the full hash:

`https://github.com/sccn/eeglab/blob/167dfc8/Contents.m#L8`

**Comment by a-hurst** [comment]
> One more general comments about the EEGLAB "translation" to Python --> i'd appreciate if you could go to the EEGLAB GitHub repo and create "permanent links" to the parts of the codebase that you used for the translation.

Do you mean per function, or per individual chunk of code (e.g. the `EI = np.ones([n_to, n_from]) - elec_dists` line you mentioned)? For the latter, it's a little messy since I refactored the code a bit for readability & consistency with MNE's versions of these functions, but I can certainly try!

File: tests/test_matprep_compare.py
Comment (a-hurst) [reply tests/test_matprep_compare.py:364]: I've updated the first comment in #73 to mention this, does that work? https://github.com/sappelhoff/pyprep/issues/73

Comment (sappelhoff) [reply tests/test_matprep_compare.py:364]: yes thank you - that's good enough!

**Comment by sappelhoff** [comment]
> Do you mean per function, or per individual chunk of code (e.g. the EI = np.ones([n_to, n_from]) - elec_dists line you mentioned)? For the latter, it's a little messy since I refactored the code a bit for readability & consistency with MNE's versions of these functions, but I can certainly try!

I don't have a system in mind, so I'd leave this up to you and recommend to do it in a "reasonable way". For example if you take something from one file in EEGLAB, then put a link ... if the file you draw from changes further down your code, put in a new link.

The readers can then be expected to follow your link and pick up the pieces themselves. that is: not every line of code must be mapped :+1:

File: docs/matlab_differences.rst
Comment (sappelhoff) [inline docs/matlab_differences.rst:216]: you may want to put a "references" heading above this one. But this is a style preference, so only do it if you agree that it's nicer.

https://sphinxcontrib-bibtex.readthedocs.io/en/latest/index.html would probably be overkill for the few citations we have, but that'd be the canonical way to solve this.

**Comment by a-hurst** [comment]
> I don't have a system in mind, so I'd leave this up to you and recommend to do it in a "reasonable way". For example if you take something from one file in EEGLAB, then put a link ... if the file you draw from changes further down your code, put in a new link.

Okay, I put line references to the definitions of the functions I reimplemented. Hopefully that's enough for anyone curious in the future!

File: docs/matlab_differences.rst
Comment (a-hurst) [reply docs/matlab_differences.rst:220]: Added it in. I think the docs look a bit better with it in there!

**Review by sappelhoff — APPROVED** [review]
thanks! :)

**Comment by sappelhoff** [comment]
@a-hurst this question on MNE discourse sounds like you could answer it: https://mne.discourse.group/t/montage-setting-differences-in-mne-vs-eeglab/3695/2

**Pull request merged**
